Add plugin creation guide to docs (#1271)
jetpack-io/devbox: docs/app/docs/guides/creating_plugins.md

## Plugin Design

Plugins are defined as Go JSON Template files, using the following schema:

```json
{
  "name": "",
  "version": "",
  "match": "",
  "readme": "",
  "env": {
    "<key>": "<value>"
  },
  "create_files": {
    "<destination>": "<source>"
  },
  "init_hook": [
    "<bash commands>"
  ]
}
```

A plugin can define services by adding a `process-compose.yaml` file in its `create_files` stanza.

### Plugin Lifecycle

Plugins are activated whenever a developer runs `devbox shell`, runs a script with `devbox run`, or starts a service using `devbox services start|restart`. The lifecycle of a devbox shell with plugins works as follows:

```mermaid
---
title: Devbox Shell Lifecycle
---
flowchart TD
   A[Plugin env] --> B
   B[User env] --> C
   C[Plugin init_hook] --> D[User Init Hook]
   D -->  E{Start Shell}
   E --> F & G & H
   F[Interactive Shell]
   G[Run Scripts]
   H[Start Services]
```

### Template Placeholders

Devbox's Plugin System provides a few special placeholders that should be used when specifying paths for env variables and helper files:

> * `{{ .DevboxDirRoot }}` – replaced with the root folder of their project, where the user's `devbox.json` is stored.

Reviewer: ```suggestion
* `{{ .DevboxDirRoot }}` – points to the root folder of their project, where the user's `devbox.json` is stored.
```

> * `{{ .DevboxDir }}` – replaced with `{{ .DevboxDirRoot }}/devbox.d/{{ plugin.name }}`. This directory is public and added to source control by default. This directory is not modified or recreated by Devbox after the initial package installation. You should use this location for files that a user will want to modify and check-in to source control alongside their project (e.g., `.conf` files or other configs).

Reviewer: ```suggestion
* `{{ .DevboxDir }}` – points to `<projectDir>/devbox.d/<plugin.name>`. This directory is public and added to source control by default. This directory is not modified or recreated by Devbox after the initial package installation. You should use this location for files that a user will want to modify and check-in to source control alongside their project (e.g., `.conf` files or other configs).
```

The `replaced with {{. }}` gets me confused

> * `{{ .Virtenv }}` – replaced with `{{ .DevboxDirRoot }}/.devbox/virtenv/{{ plugin.name }}` whenever the plugin activates. This directory is hidden and added to `.gitignore` by default You should use this location for files or variables that a user should not check-in or edit directly. Files in this directory should be considered managed by Devbox, and may be recreated or modified after the initial installation.

Reviewer: ditto

### Fields

#### `name` *string*

The name of your plugin. This is used to identify your plugin when a user runs `devbox info`. If `match` is not set, the plugin will automatically activate when a package is added to a devbox.json project that matches `name`.

#### `version` *string*

The version of your plugin. You should kstart your version at 0.0.1 and bump it whenever you merge an update to the plugin.

#### `match` *string*

A regex expression that is used to identify when the plugin will be activated. Devbox will activate your plugin when a package installed with `devbox add` matches this regular expression.

> The regex you provide should match a plugin name in the `nixpkgs` repository. You can look up packages at `search.nixos.org`

Reviewer: Match a plugin name?? Or should it be matching a package name in `nixhub.io` directly?
